app/src/main/java/com/example/secret/PostsListFragment.java

```diff
@@ -0,0 +1,85 @@
+import com.example.secret.model.PostsModel;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class PostsListFragment extends Fragment {
+    FragmentPostsListBinding binding;
+    PostRecyclerAdapter adapter;
+    PostsListFragmentViewModel viewModel;
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        // Inflate the layout for this fragment
+        binding = FragmentPostsListBinding.inflate(inflater, container, false);
+        View view = binding.getRoot();
+
+        binding.recyclerView.setHasFixedSize(true);
+        binding.recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
+        Map<String, List<Comment>> postsLatestCommentsData = new HashMap<>();
+        for (Map.Entry<String, LiveData<List<Comment>>> entry : viewModel.getPostsLatestComments().entrySet()) {
+            postsLatestCommentsData.put(entry.getKey(), entry.getValue().getValue());
+        }
+        adapter = new PostRecyclerAdapter(getLayoutInflater(), viewModel.getPosts().getValue(), postsLatestCommentsData);
+        binding.recyclerView.setAdapter(adapter);
+
+//        adapter.setOnItemClickListener(pos -> {
+//            Log.d("TAG", "Row was clicked " + pos);
+//            Post st = viewModel.getData().getValue().get(pos);
+////                NavDirections.ActionPostsListFragmentToBlueFragment action = PostsListFragmentDirections.actionPostsListFragmentToBlueFragment(st.name);
+////                Navigation.findNavController(view).navigate(action);
+//        });
+
+        binding.progressBar.setVisibility(View.GONE);
+
+        viewModel.getPosts().observe(getViewLifecycleOwner(), postsList -> {
+            adapter.setPosts(postsList);
+            for (Post post : postsList) {
+                viewModel.getPostLatestComments(post.id).observe(getViewLifecycleOwner(),
+                        commentList -> adapter.setPostLatestComments(post.id, commentList));
+            }
+        });
+
+        PostsModel.instance().EventPostsListLoadingState.observe(getViewLifecycleOwner(), status ->
+                binding.swipeRefresh.setRefreshing(status == PostsModel.LoadingState.LOADING));
+
+        CommentsModel.instance().eventCommentsListLoadingState.observe(getViewLifecycleOwner(), status ->
+                binding.swipeRefresh.setRefreshing(status == CommentsModel.LoadingState.LOADING));
+
+        binding.swipeRefresh.setOnRefreshListener(this::reloadData);
+        return view;
+    }
+
+    @Override
+    public void onAttach(@NonNull Context context) {
+        super.onAttach(context);
+        viewModel = new ViewModelProvider(this).get(PostsListFragmentViewModel.class);
+    }
+
+    void reloadData() {
+        binding.progressBar.setVisibility(View.VISIBLE);
+        PostsModel.instance().refreshLatestPosts();
+        CommentsModel.instance().refreshLatestComments();
```

Comment:
you are invoking two async methods here. setting the visibility after the invocation will misguide the user.
Instead, set listeners to check when the process is actually finished

Reply:
No need to set the progress bar to visable from the first place because the refresh view handles that.

Reply:
Pushed a fix